content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md

```diff
@@ -0,0 +1,128 @@
+and therefore saves unnecessary scheduling attempts. Those Pods will also be ignored by Cluster
+Autoscaler if you have it installed in the cluster.
+
+Clearing the gates is the responsibility of external controllers with knowledge of when the Pod
+should be considered for scheduling (e.g., a quota manager).
+
+{{< mermaid >}}
+graph LR;
+  pod((New Pod))-->queue
+  subgraph Scheduler
+    queue(scheduler queue)
+    sched_cycle[/scheduling cycle/]
+    schedulable{schedulable?}
+    popout{Pop out?}
+    
+    queue==>|PreEnqueue check|popout
+    popout-->|Yes|sched_cycle
+    popout==>|No|queue
+    sched_cycle-->schedulable
+    schedulable-->|No|queue
+    subgraph note [A knob to gate Pod's scheduling]
+    end
+  end
+  
+ classDef plain fill:#ddd,stroke:#fff,stroke-width:1px,color:#000;
+ classDef k8s fill:#326ce5,stroke:#fff,stroke-width:1px,color:#fff;
+ classDef Scheduler fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
+ classDef note fill:#edf2ae,stroke:#fff,stroke-width:1px;
+ classDef popout fill:#f96,stroke:#fff,stroke-width:1px;
+ class queue,sched_cycle,schedulable k8s;
+ class pod plain;
+ class note note;
+ class popout popout;
+ class Scheduler Scheduler;
+{{< /mermaid >}}
+
+## How does it work?
+
+Scheduling gates in general works very similar to Finalizers. Pods with non-empty 
+`spec.schedulingGates` field will show as status `SchedulingGated` and be blocked from
+scheduling. Note that more than one gate can be added, but they all should be added upon Pod
+creation (e.g., you can add them as part of the spec or via a mutating webhook).
+
+```
+NAME       READY   STATUS            RESTARTS   AGE
+test-pod   0/1     SchedulingGated   0          10s
+```
+
+To clear the gates, you update the Pod by removing all of the items from the Pod's `schedulingGates`
+field. The gates do not need to be removed all at once, but only when all the gates are removed the
+scheduler will start to consider the Pod for scheduling.
+
+Under the hood, scheduling gates are implemented as a PreEnqueue scheduler plugin, a new scheduler
+framework extension point that is invoked at the beginning of each scheduling cycle.
+
+## Use Cases
+
+An important use case this feature enables is dynamic quota management. Kubernetes supports
+[ResourceQuota](/docs/concepts/policy/resource-quotas/), however the API Server enforces quota at
+the time you attempt Pod creation. For example, if a new Pod exceeds the CPU quota, it gets rejected.
+The API Server doesn't queue the Pod; therefore, whoever created the Pod needs to continuously attempt
+to recreate it again. This either means a delay between resources becoming available and the Pod
+actually running, or it means load on the API server and Scheduler due to constant attempts.
+
+Scheduling gates allows an external quota manager to address the above limitation of ResourceQuota.
+Specifically, the manager could add a `example-com/quota-check` scheduling gate to all Pods created in the
```

Comment:
```suggestion
Specifically, the manager could add a `example.com/quota-check` scheduling gate to all Pods created in the
```
If `example.com/quota-check` isn't a valid scheduling gate name, is that actually something we'd want to fix for beta?

I'd expect an analogy with https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set and for similar motivations: different products that people might install onto a cluster can avoid clashes between keys.

Reply:
For alpha impl., we don't enforce the name to be full qualified style. It's tracked for beta to evaluate: [#113608](https://github.com/kubernetes/kubernetes/issues/113608)

Reply:
`-` was a typo... it's updated with `example.com/quota-check`.

Reply:
Ah! All is clear. Thanks for explaining.